boards/shields/wiznet_wiz550io/wiznet_wiz550io.c

```diff
@@ -0,0 +1,157 @@
+ * mac address configured.
+ */
+#define WIZ550IO_USE_MAC_ADDRESS                                                                   \
+	(!DT_PROP(W5500_NODE, zephyr_random_mac_address) &&                                        \
+	 !DT_NODE_HAS_PROP(W5500_NODE, local_mac_address))
+
+/* W5500 registers */
+#define W5500_SHAR 0x0009 /* Source MAC address */
+
+#define W5500_SPI_BLOCK_SELECT(addr) (((addr) >> 16) & 0x1f)
+#define W5500_SPI_READ_CONTROL(addr) (W5500_SPI_BLOCK_SELECT(addr) << 3)
+
+/* WIZ550io documentation recommends 150ms delay after HW reset
+ * for PIC12F519 MCU to configure the W5500.
+ */
+#define WIZ550IO_RESET_DELAY K_MSEC(150)
+
+static struct net_eth_addr wiz550io_mac_addr;
+
+static int wiz550io_spi_read_mac(struct net_eth_addr *mac_addr)
+{
+	struct spi_dt_spec spi = SPI_DT_SPEC_GET(W5500_NODE, SPI_WORD_SET(8), 0);
+	const uint32_t addr = W5500_SHAR;
+	int ret;
+
+	if (!spi_is_ready_dt(&spi)) {
+		LOG_ERR("SPI %s not ready", spi.bus->name);
+		return -EINVAL;
+	}
+
+	uint8_t cmd[3] = {addr >> 8, addr, W5500_SPI_READ_CONTROL(addr)};
+	const struct spi_buf tx_buf = {
+		.buf = cmd,
+		.len = ARRAY_SIZE(cmd),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	/* skip the default dummy 0x010203 */
+	const struct spi_buf rx_buf[2] = {
+		{.buf = NULL, .len = 3},
+		{.buf = mac_addr->addr, .len = sizeof(mac_addr->addr)},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	ret = spi_transceive_dt(&spi, &tx, &rx);
+
+	return ret;
+}
+
+static int wiz550io_pre_init(void)
+{
+	int ret;
+
+	/* Hold reset pin low for 500us to perform a hardware reset */
+	gpio_pin_set(WIZ550IO_RESET_PORT, WIZ550IO_RESET_PIN, 1);
+	k_usleep(500);
+	gpio_pin_set(WIZ550IO_RESET_PORT, WIZ550IO_RESET_PIN, 0);
+
+	/* Wait for the device to be ready */
+	if (WIZ550IO_HAS_READY_GPIO) {
+		k_timepoint_t timeout = sys_timepoint_calc(WIZ550IO_RESET_DELAY);
+
+		while (!gpio_pin_get(WIZ550IO_READY_PORT, WIZ550IO_READY_PIN)) {
+			if (sys_timepoint_expired(timeout)) {
+				LOG_ERR("WIZ550io not ready");
+				return -ETIMEDOUT;
+			}
+			k_msleep(1);
+		}
+	} else {
+		k_sleep(WIZ550IO_RESET_DELAY);
+	}
+
+	if (WIZ550IO_USE_MAC_ADDRESS) {
+		/* Read the WIZ550io unique MAC address set by the PIC12F519.
+		 * It needs to be read after a hardware reset and before software reset.
+		 */
+		ret = wiz550io_spi_read_mac(&wiz550io_mac_addr);
+		if (ret) {
+			LOG_ERR("WIZ550io unable to read MAC address");
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static int wiz550io_post_init(void)
+{
+	int ret = 0;
+
+	if (WIZ550IO_USE_MAC_ADDRESS) {
+		/* Assign the WIZ550io unique MAC address */
+		const struct device *w5500_dev = DEVICE_DT_GET(W5500_NODE);
+		const struct ethernet_api *w5500_api = w5500_dev->api;
+		struct ethernet_config config = {.mac_address = wiz550io_mac_addr};
+
+		ret = w5500_api->set_config(w5500_dev, ETHERNET_CONFIG_TYPE_MAC_ADDRESS, &config);
+	}
+
+	return ret;
+}
```

Comment:
This should be moved to the w5500 device driver, implemented in the `get_config()` function of the `ethernet_api`.

This code is essentially an ad-hoc implementation of part of the `get_config()` API, which is then used to overwrite the config using `set_config()`

Reply:
See
https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/drivers/ethernet/eth_w5500.c#L462-L469
and
https://github.com/zephyrproject-rtos/zephyr/blob/9b5260de9a31738f5632a0f7240f1f70bc1d4cd4/include/zephyr/net/ethernet.h#L557-L560

Reply:
@bjarki-andreasen I'm not sure I understand, could you elaborate?

It looks like `get_config` is used by `net_eth_get_hw_config` to "Return ethernet device hardware configuration information". How does this help to set the mac address?

Reply:
Strictly speaking, you could argue this breaks encapsulation, but I think it's fine since this board is basically just the w5500, there's not much to be gained from encapsulation. And chucking board specific code into the ethernet mac driver any time somebody introduces a breakout board with this mac doesn't make any sense. That would quickly get out of hand. 

I have absolutely no idea why so many people are complaining about breaking the abstraction of the ethernet api by calling a driver function from the (known) device struct while simultaneously suggesting to put a bunch of board specific code in the MAC driver, which is an even worse breach of abstraction, IMO. Can somebody explain to me the thought process there instead of just saying to do these things, we already talked a lot about this on this PR, and let's stop going back and forth on things.

Reply:
The MAC address is read from the WF5500 chip AFAIK, therefore this is not violating any encapsulation. All the external chip does is set the MAC address using the same interface as `set_config()` does. No board specific code is going into the MAC driver, but MAC specific code is moved to the MAC driver from the board code.

Reply:
@gramsay0 The MAC address is one of those hardware configurations :) So when a higher level subsystem "probes" the device using `get_config()`, it will retrieve and set the MAC address which has been assigned to the WF5500 by the external chip.

Reply:
@bjarki-andreasen ahh thanks I understand now.

The problem is reset sequencing. A hard (GPIO) reset is required for the external chip to write the MAC address into W5500. But then the soft (SPI command) reset in W5500 init will clear it.

Earlier we discussed disabling the soft reset in W5500 init, although it's unclear exactly which registers the external chip modifes. It seemed safer to keep the soft reset to put the device in a known state, the unique MAC is all we are interested in

Reply:
You do care about the MAC address because it is an OUI number for which you have to buy an identifier and have a block of addresses to assign to devices (with one unique number per device). You can't just give devices random addresses, despite some manufacturers being nonchalant on the matter, it is not legal